File: src/main/scala/mmlTests/a201504/scala/gpt35turbo/TreeImpl.scala
package mmlTests.a201504.scala.gpt35turbo

import scala.annotation.tailrec
import scala.collection.mutable.ListBuffer
import mmlTests.a201504.scala.Tree

class TreeImpl[X](val root: X, val sons: List[Tree[X]]) extends Tree[X] {
  override def size: Int = {
    @tailrec
    def sizeHelper(remaining: List[Tree[X]], acc: Int): Int = {
      remaining match {
        case Nil => acc
        case head :: tail => sizeHelper(tail, acc + head.size)
      }
    }

    sizeHelper(sons, 1)
  }

  override def getRoot: X = root

  override def getSons: List[Tree[X]] = sons

  override def contains(x: X): Boolean = {
    @tailrec
    def containsHelper(remaining: List[Tree[X]]): Boolean = {
      remaining match {
        case Nil => false
        case head :: tail =>
          if (head.getRoot == x || head.contains(x)) true
          else containsHelper(tail)
      }
    }

    containsHelper(sons)
  }

  override def getSubTree(node: X): Tree[X] = {
    @tailrec
    def getSubTreeHelper(remaining: List[Tree[X]]): Tree[X] = {
      remaining match {
        case Nil => null
        case head :: tail =>
          if (head.getRoot == node) head
          else {
            val subTree = head.getSubTree(node)
            if (subTree != null) return subTree
            getSubTreeHelper(tail)
          }
      }
    }

    getSubTreeHelper(sons)
  }

  override def toList: List[X] = {
    @tailrec
    def toListHelper(remaining: List[Tree[X]], acc: List[X]): List[X] = {
      remaining match {
        case Nil => acc
        case head :: tail => toListHelper(tail, acc ::: head.toList)
      }
    }

    root :: toListHelper(sons, Nil)
  }
}

File: src/main/scala/mmlTests/a201504/java/gpt35turbo/TreeImpl.java
package mmlTests.a201504.java.gpt35turbo;

import mmlTests.a201504.java.Tree;
import mmlTests.a201504.java.TreeFactory;
import java.util.ArrayList;
import java.util.List;

public class TreeImpl<X> implements Tree<X> {

    private X root;
    private List<Tree<X>> sons;

    public TreeImpl(X root, List<Tree<X>> sons) {
        this.root = root;
        this.sons = sons;
    }

    @Override
    public int size() {
        int size = 1;
        for (Tree<X> son : sons) {
            size += son.size();
        }
        return size;
    }

    @Override
    public X getRoot() {
        return root;
    }

    @Override
    public List<Tree<X>> getSons() {
        return new ArrayList<Tree<X>>(sons);
    }

    @Override
    public boolean contains(X x) {
        if (root.equals(x)) {
            return true;
        }
        for (Tree<X> son : sons) {
            if (son.contains(x)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public Tree<X> getSubTree(X node) {
        if (root.equals(node)) {
            return this;
        }
        for (Tree<X> son : sons) {
            Tree<X> subTree = son.getSubTree(node);
            if (subTree != null) {
                return subTree;
            }
        }
        return null;
    }

    @Override
    public List<X> toList() {
        List<X> list = new ArrayList<X>();
        list.add(root);
        for (Tree<X> son : sons) {
            list.addAll(son.toList());
        }
        return list;
    }

}


